### PR TITLE
JanusGraph examples

### LINKED ISSUE #530: JanusGraph examples
It would help to on-board new users with example code on how to write an application that uses JanusGraph with the various storage and indexing backend configurations. `GraphOfTheGods` does this in part, but it isn't a standalone code example that a developer could easily clone and replace the logic to build and query their own.

### PR DESCRIPTION
**Pull request by pluradj**
Fixes #530

Signed-off-by: Jason Plurad <pluradj@us.ibm.com>

Thank you for contributing to JanusGraph.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there an issue associated with this PR? Is it referenced in the commit message?
- [X] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [X] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you written and/or updated unit tests to verify your changes?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [X] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [X] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-examples/example-remotegraph/conf/remote-objects.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+hosts: [127.0.0.1]
+port: 8182
+serializer: {
```
Comment (mbrukman) [inline janusgraph-examples/example-remotegraph/conf/remote-objects.yaml]: This is mixing JSON in a YAML config, no? I understand that JSON is a subset of YAML, but this can also be written as:

```yaml
serializer:
  className: org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV1d0
    config:
      ioRegistries: [org.janusgraph.graphdb.tinkerpop.JanusGraphIoRegistry]
```

Or would that not do the same thing?

File: janusgraph-examples/example-cassandra/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+The keyspace name matches the value of `storage.cassandra.keyspace`
+
+```
+$ cd $CASSANDRA_HOME
+$ bin/cqlsh -e 'describe keyspaces'
+system_traces  system  jgex
+$ bin/cqlsh -e 'drop keyspace jgex'
+$ bin/cqlsh -e 'describe keyspaces'
+system_traces  system
+```
```
Comment (sjudeng) [inline janusgraph-examples/example-cassandra/README.md]: You could also use the core factory API (`JanusGraphFactory.drop(Graph)`) to drop the keyspace. The drop method is fully supported on all storage backends (Cassandra/CQL, HBase, BerkeleJE) and indexing backends with the exception of Solr (HTTP client).

Comment (pluradj) [reply janusgraph-examples/example-cassandra/README.md]: Good call, I don't think that was implemented yet when I first started working on these examples.

Comment (sjudeng) [reply janusgraph-examples/example-cassandra/README.md]: Note it currently throws an exception when using Solr in HTTP mode but there's a proposal to lower that to an error message in #534. Either way the storage backend is still dropped.

File: janusgraph-examples/example-remotegraph/conf/remote-objects.yaml (hunk 1 above)
Comment (pluradj) [reply janusgraph-examples/example-remotegraph/conf/remote-objects.yaml]: Using curly braces is the same style used in the reference TinkerPop [remote-objects.yaml](https://github.com/apache/tinkerpop/blob/3.2.6/gremlin-console/conf/remote-objects.yaml).

Comment (pluradj) [reply janusgraph-examples/example-remotegraph/conf/remote-objects.yaml]: And it appears that the Java parser wants those curly braces in place otherwise it fails to parse the flow mapping.

**Comment by pluradj** [comment]
@sjudeng @mbrukman thanks for all the review comments. I pushed fixes for everything suggested.

File: janusgraph-examples/example-remotegraph/conf/remote-objects.yaml (hunk 1 above)
Comment (robertdale) [reply janusgraph-examples/example-remotegraph/conf/remote-objects.yaml]: I am in favor of keeping in sync with TinkerPop style. However, for education purposes only, the correct pure YAML format is (and works with TinkerPop):
```
hosts: 
  - localhost
port: 8182
serializer: 
  className: org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV1d0
  config: 
    ioRegistries: 
      - org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV1d0
```

File: janusgraph-examples/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# JanusGraph Examples
+
+The JanusGraph examples show the basics of how to configure and construct
+a graph application. It uses [Apache Maven](https://maven.apache.org) to
+manage the numerous dependencies required to build the application. The common
+application:
```
Comment (mbrukman) [inline janusgraph-examples/README.md]: The common application **will**:

File: janusgraph-examples/example-berkeleyje/README.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+for additional properties.
+
+## Running the example
+
+Use [Apache Maven](http://maven.apache.org/) and the
+[exec-maven-plugin](http://www.mojohaus.org/exec-maven-plugin/java-mojo.html)
+to pull in the required jar files onto the runtime classpath.
+
+```
+# Run from $JANUSGRAPH_HOME/janusgraph-examples/example-berkeleyje
```
Comment (mbrukman) [inline janusgraph-examples/example-berkeleyje/README.md]: Make this explicit? I.e., change from a comment to:

```
cd $JANUSGRAPH_HOME/janusgraph-examples/example-berkeleyje
```

Similarly elsewhere.

File: janusgraph-examples/example-remotegraph/conf/remote-objects.yaml (hunk 1 above)
Comment (mbrukman) [reply janusgraph-examples/example-remotegraph/conf/remote-objects.yaml]: LGTM

File: janusgraph-examples/example-tinkergraph/conf/jgex-tinkergraph.properties [hunk 5]
```diff
@@ -0,0 +1,6 @@
+gremlin.graph=org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerGraph
+
+// DefaultIdManager types: INTEGER, LONG, UUID, ANY
```
Comment (mbrukman) [inline janusgraph-examples/example-tinkergraph/conf/jgex-tinkergraph.properties]: The comment characters in a properties file is `#`, not `//`:

https://docs.oracle.com/cd/E23095_01/Platform.93/ATGProgGuide/html/s0204propertiesfileformat01.html

Incidentally, `!` appears to also be recognized as a comment character, but I would suggest sticking with the more standard `#` since that's more widely used in many other file formats.

**Comment by sjudeng** [comment]
@pluradj I was in the middle of a review here. Can I still submit maybe for a follow-on PR?

File: janusgraph-examples/example-hbase/README.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+$ server/scripts/cloud-scripts/zkcli.sh -z 127.0.0.1:9983 -cmd clear /configs/jgex
+```
+
+
+## HBase and Solr HTTP
+
+### Create the Solr cores
+
+The core names match the `vAge` or `eReasonPlace` values when `mgmt.buildIndex()`
+defines the mixed indexes in `JanusGraphApp.createMixedIndexes()`
```
Comment (sjudeng) [inline janusgraph-examples/example-hbase/README.md]: Can you rephrase?

File: janusgraph-examples/example-hbase/README.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+```
+$ cd $JANUSGRAPH_HOME/janusgraph-examples/example-hbase
+
+$ mvn exec:java -Dexec.mainClass="org.janusgraph.example.JanusGraphApp" -Dlogback.configurationFile="conf/logback.xml" -Dexec.args="conf/jgex-hbase-solr-http.properties drop"
+```
+
+### Drop the Solr cores
+
+The core names match the `vAge` or `eReasonPlace` values when `mgmt.buildIndex()`
+defines the mixed indexes in `JanusGraphApp.createMixedIndexes()`
```
Comment (sjudeng) [inline janusgraph-examples/example-hbase/README.md]: Rephrase

File: janusgraph-examples/example-remotegraph/README.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+client. Sumbit scripts as a `String` for evaluation on the server.
+* [Remote Graph](http://tinkerpop.apache.org/docs/3.2.6/reference/#connecting-via-remotegraph)
+uses the same client connection but enables you to compose queries as code.
+
+## JanusGraph configuration
+
+### Gremlin Server configuration
+
+* Select a specific storage backend to use, then copy its configuration into
+the `$JANUSGRAPH_HOME/conf/gremlin-server/` directory
```
Comment (sjudeng) [inline janusgraph-examples/example-remotegraph/README.md]: Is this still in the context of a source JanusGraph repository or is it a binary distribution? If source, same comment as above regarding the presence of `conf` in root directory here and below. If binary, make a note that a separate binary distribution is assumed for managing the Gremlin Server instance.

File: janusgraph-examples/example-remotegraph/README.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+* Update the `$JANUSGRAPH_HOME/conf/gremlin-server/gremlin-server.yaml` to
+use the configuration file
```
Comment (sjudeng) [inline janusgraph-examples/example-remotegraph/README.md]: ", for example"

File: janusgraph-examples/example-remotegraph/README.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+allows a JanusGraph instance to run on a standalone server. This enables
+multiple clients to connect to the same JanusGraph instance. It also allows
+clients to connect via non-JVM languages. By default, the Gremlin Server
+communicates over WebSocket.
+
+A Java program can connect to the Gremlin Server in two ways:
+* [Gremlin Driver](http://tinkerpop.apache.org/docs/3.2.6/reference/#connecting-via-java)
+client. Sumbit scripts as a `String` for evaluation on the server.
+* [Remote Graph](http://tinkerpop.apache.org/docs/3.2.6/reference/#connecting-via-remotegraph)
+uses the same client connection but enables you to compose queries as code.
```
Comment (sjudeng) [inline janusgraph-examples/example-remotegraph/README.md]: It might be easy to forget to update the TinkerPop link versions. Is this common throughout the docs and/or do we need an issue to track adding support for automatically updating version when we upgrade TinkerPop?

File: janusgraph-examples/example-cassandra/pom.xml [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.janusgraph</groupId>
+        <artifactId>janusgraph-examples</artifactId>
+        <version>0.2.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <artifactId>example-cassandra</artifactId>
+    <packaging>pom</packaging>
+    <name>Example-Cassandra: C* Thrift Storage, ES Index</name>
```
Comment (sjudeng) [inline janusgraph-examples/example-cassandra/pom.xml]: `/ES/Elasticsearch/`

File: janusgraph-examples/example-cql/pom.xml [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.janusgraph</groupId>
+        <artifactId>janusgraph-examples</artifactId>
+        <version>0.2.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <artifactId>example-cql</artifactId>
+    <packaging>pom</packaging>
+    <name>Example-Cql: C* CQL Storage, ES Index</name>
```
Comment (sjudeng) [inline janusgraph-examples/example-cql/pom.xml]: `/ES/Elasticsearch/`

File: janusgraph-examples/example-common/src/main/java/org/janusgraph/example/GraphApp.java [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+            Vertex sea = g.addV("location").property("name", "sea").next();
+            Vertex jupiter = g.addV("god").property("name", "jupiter").property("age", 5000).next();
+            Vertex neptune = g.addV("god").property("name", "neptune").property("age", 4500).next();
+            Vertex hercules = g.addV("demigod").property("name", "hercules").property("age", 30).next();
+            Vertex alcmene = g.addV("human").property("name", "alcmene").property("age", 45).next();
+            Vertex pluto = g.addV("god").property("name", "pluto").property("age", 4000).next();
+            Vertex nemean = g.addV("monster").property("name", "nemean").next();
+            Vertex hydra = g.addV("monster").property("name", "hydra").next();
+            Vertex cerberus = g.addV("monster").property("name", "cerberus").next();
+            Vertex tartarus = g.addV("location").property("name", "tartarus").next();
```
Comment (sjudeng) [inline janusgraph-examples/example-common/src/main/java/org/janusgraph/example/GraphApp.java]: `final` on all above

File: janusgraph-examples/example-common/src/main/java/org/janusgraph/example/GraphApp.java [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+    public void readElements() {
+        try {
+            if (g == null) {
+                return;
+            }
+
+            LOGGER.info("reading elements");
+
+            // look up vertex by name can use a composite index in JanusGraph
+            Optional<Map<String, Object>> v = g.V().has("name", "jupiter").valueMap(true).tryNext();
```
Comment (sjudeng) [inline janusgraph-examples/example-common/src/main/java/org/janusgraph/example/GraphApp.java]: `final` here and throughout

File: janusgraph-examples/README.md [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+* Run traversal queries to get data from the graph
+* Make updates to the graph
+* Close the graph
+
+By using different graph configurations, the same example code can run against
+the various supported storage and indexing backends.
+
+## Prerequisites
+
+* Java 8 Developer Kit, update 40 or higher
```
[the remark this answers is not in the document]
Comment (pluradj) [reply janusgraph-examples/README.md]: I don't think it matters, so I didn't specify. I've used both successfully.

File: janusgraph-examples/example-berkeleyje/README.md [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+## Drop the graph
+
+Make sure to stop the application before dropping the graph. The configuration
+uses the application name `jgex` as the root directory for the BerkeleyJE
+and Lucene directories.
+
+```
+$ cd $JANUSGRAPH_HOME/janusgraph-examples/example-berkeleyje
+
+$ mvn exec:java -Dexec.mainClass="org.janusgraph.example.JanusGraphApp" -Dexec.args="conf/jgex-berkeleyje.properties drop"
```
[the remark this answers is not in the document]
Comment (pluradj) [reply janusgraph-examples/example-berkeleyje/README.md]: I chose to include `$` for consistency with places where I chose to show the output of those commands. Understand on the cut & paste aspect though.

File: janusgraph-examples/example-remotegraph/README.md (hunk 10 above)
Comment (pluradj) [reply janusgraph-examples/example-remotegraph/README.md]: Good call, there's a syntax to use where the value can be replaced with the Maven value. I'll update this.

File: janusgraph-examples/example-cassandra/pom.xml (hunk 11 above)
Comment (pluradj) [reply janusgraph-examples/example-cassandra/pom.xml]: I can do this (expand C* also), but I was trying to keep the formatting of the Maven build completion message to fit nicely ;)

File: janusgraph-examples/example-hbase/README.md [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+
+Using a configset makes it possible to reuse the same configuration for new
+cores. The configset is stored in Zookeeper under `/configs/jgex` where the
+name `jgex` matches the properties file value for `index.jgex.solr.configset`.
+Make sure the Zookeeper url matches the properties value for `index.jgex.solr.zookeeper-url`.
+
+```
+$ cd $SOLR_HOME
+
+$ server/scripts/cloud-scripts/zkcli.sh -z 127.0.0.1:9983 -cmd upconfig -d $JANUSGRAPH_HOME/conf/solr -n jgex
```
[the remark this answers is not in the document]
Comment (pluradj) [reply janusgraph-examples/example-hbase/README.md]: The `$JANUSGRAPH_HOME` references here are more in context of working with the `examples` packaged in the distribution zip (i.e. `JANUSGRAPH_HOME=/usr/lib/janusgraph-0.2.0-hadoop2`), not from a cloned source code repository.

File: janusgraph-examples/example-remotegraph/README.md (hunk 8 above)
Comment (pluradj) [reply janusgraph-examples/example-remotegraph/README.md]: The context for all the examples is the binary distribution. The examples are packaged with the distribution so users can download the distribution then build the examples.

File: janusgraph-examples/example-hbase/README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (sjudeng) [reply janusgraph-examples/example-hbase/README.md]: @pluradj Makes sense. Can you make a note of this somewhere up front (if you haven't already)?
